File: Februrary 2025/01.cpp
class Solution {
  public:
    bool isWordExist(vector<vector<char>>& mat, string& word) {
        int n = mat.size();
        int m = mat[0].size();
        vector<vector<bool>> visited(n, vector<bool>(m, false));

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (mat[i][j] == word[0]) {
                    if (dfs(mat, i, j, word, 0, visited)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

  private:
    int rowDir[4] = {0, 0, -1, 1}; // Left, Right, Up, Down
    int colDir[4] = {-1, 1, 0, 0};

    bool dfs(vector<vector<char>>& mat, int row, int col, string& word, int index, vector<vector<bool>>& visited) {
        if (index == word.size()) {
            return true;
        }

        if (row < 0 || col < 0 || row >= mat.size() || col >= mat[0].size() || visited[row][col] || mat[row][col] != word[index]) {
            return false;
        }

        visited[row][col] = true;

        for (int d = 0; d < 4; d++) {
            int newRow = row + rowDir[d];
            int newCol = col + colDir[d];

            if (dfs(mat, newRow, newCol, word, index + 1, visited)) {
                return true;
            }
        }

        visited[row][col] = false;
        return false;
    }
};

File: Februrary 2025/01.py
class Solution:
    def __init__(self):
        self.rowDir = [0, 0, -1, 1] # Left, Right, Up, Down
        self.colDir = [-1, 1, 0, 0]

    def isWordExist(self, mat, word):
        n = len(mat)
        m = len(mat[0])

        # Iterate through the matrix to find the first character
        for i in range(n):
            for j in range(m):
                if mat[i][j] == word[0]:
                    if self.dfs(mat, i, j, word, 0, [[False]*m for _ in range(n)]):
                        return True
        return False

    def dfs(self, mat, row, col, word, index, visited):
        if index == len(word):
            return True

        # Boundary conditions and character match check
        if row < 0 or col < 0 or row >= len(mat) or col >= len(mat[0]) or visited[row][col] or mat[row][col] != word[index]:
            return False

        # Mark as visited
        visited[row][col] = True

        # Explore all 4 directions
        for d in range(4):
            newRow = row + self.rowDir[d]
            newCol = col + self.colDir[d]

            if self.dfs(mat, newRow, newCol, word, index + 1, visited):
                return True

        # Backtrack
        visited[row][col] = False
        return False
